flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -174,7 +174,9 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
 			LOG.debug("Notification of complete checkpoint for task {}", taskName);
 
 			for (StreamOperatorWrapper<?, ?> operatorWrapper : operatorChain.getAllOperators(true)) {
-				operatorWrapper.getStreamOperator().notifyCheckpointComplete(checkpointId);
+				if (!operatorWrapper.isClosed()) {
```

Comment:
I wonder if we shouldn't throw some exception?

Reply:
It depends on what we are trying to achieve. Assume that we have a chain source -> map -> sink and source and map have been closed already. Wouldn't it be nice to still commit stuff in sink?

Reply:
If one operator is closed in the chain, doesn't it mean all of them are closed?

Reply:
Not with the wrapper. It closes them one after another from head to tail with mails. That's why the notify can sneak in in the first place.